File: tools/wizard/modules/wdg.lua
--[[============================================================================
@file    eth.lua

@author  Daniel Zorychta

@brief   Ethernet configuration wizard.

@note    Copyright (C) 2014 Daniel Zorychta <daniel.zorychta@gmail.com>

         This program is free software; you can redistribute it and/or modify
         it under the terms of the GNU General Public License as published by
         the  Free Software  Foundation;  either version 2 of the License, or
         any later version.

         This  program  is  distributed  in the hope that  it will be useful,
         but  WITHOUT  ANY  WARRANTY;  without  even  the implied warranty of
         MERCHANTABILITY  or  FITNESS  FOR  A  PARTICULAR  PURPOSE.  See  the
         GNU General Public License for more details.

         You  should  have received a copy  of the GNU General Public License
         along  with  this  program;  if not,  write  to  the  Free  Software
         Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


*//*========================================================================--]]

require "defs"
require "db"
require "pll"

--------------------------------------------------------------------------------
-- OBJECTS
--------------------------------------------------------------------------------
wdg = {}

local arch = {}
arch.stm32f1            = {}
arch.stm32f1.configure  = nil

arch.stm32f2            = {}
arch.stm32f2.configure  = nil

arch.stm32f3            = {}
arch.stm32f3.configure  = nil

arch.stm32f4            = {}
arch.stm32f4.configure  = nil

--------------------------------------------------------------------------------
-- FUNCTIONS
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
-- @brief Calculate total steps
--------------------------------------------------------------------------------
local function calculate_total_steps()
        local arch   = db:get_arch()
        local family = db:get_mcu_family(db:get_cpu_name())

        if key_read(db.path.project.mk, "ENABLE_ETH") == yes then
                if arch == "stm32f1" then
                        progress(1, 4)
                elseif arch == "stm32f2" then
                elseif arch == "stm32f3" then
                elseif arch == "stm32f4" then
                end
        else
                progress(1, 1)
        end
end

--------------------------------------------------------------------------------
-- @brief Ask user to select enable/disable module
--------------------------------------------------------------------------------
local function ask_for_enable()
        local choice = key_read(db.path.project.flags, "__ENABLE_WDG__")
        msg(progress().."Do you want to enable WDG module?")
        msg("Current selection is: "..filter_yes_no(choice)..".")
        add_item(yes, "Yes")
        add_item(no, "No")
        choice = get_selection()
        if can_be_saved(choice) then
                key_save(db.path.project.flags, "__ENABLE_WDG__", choice)
                key_save(db.path.project.mk, "ENABLE_WDG", choice)
        end

        calculate_total_steps()
        progress(2)

        return choice
end

--------------------------------------------------------------------------------
-- @brief Configuration for STM32F1
--------------------------------------------------------------------------------
arch.stm32f1.configure = function()

        local function configure_dev_lock_at_open()
                local choice = key_read(db.path.stm32f1.wdg.flags, "__WDG_DEVICE_LOCK_AT_OPEN__")
                msg(progress().."Do you want to lock WDG device at first open?")
                msg("Current selection is: "..filter_yes_no(choice)..".")
                add_item(yes, "Yes")
                add_item(no, "No")
                choice = get_selection()
                if can_be_saved(choice) then
                        key_save(db.path.stm32f1.wdg.flags, "__WDG_DEVICE_LOCK_AT_OPEN__", choice)
                end

                return choice
        end

        local function configure_wdg_timeout()
                local divtab  = {{25,    256, 3907},
                                 {20,    256, 3125},
                                 {15,    256, 2344},
                                 {10,    256, 1563},
                                 {5,     256, 782 },
                                 {1,     256, 157 },
                                 {0.5,   64,  313 },
                                 {0.25,  32,  313 },
                                 {0.125, 16,  313 },
                                 {0.064, 8,   320 },
                                 {0.032, 4,   320 },
                                 {0.016, 4,   160 },
                                 {0.008, 4,   80  },
                                 {0.004, 4,   40  },
                                }

                local divider = key_read(db.path.stm32f1.wdg.flags, "__WDG_CLK_DIVIDER__")
                local reload  = key_read(db.path.stm32f1.wdg.flags, "__WDG_RELOAD_VALUE__")
                local lsi     = pll:get_frequencies().LSI
                local choice  = tunit(round(1/(lsi/divider/reload), 3))
                msg(progress().."Select Watchdog timeout.")
                msg("Current selection is: "..choice..".")
                for i = 1, #divtab do
                        add_item(i, tunit(divtab[i][1]))
                end
                choice = get_selection()
                if can_be_saved(choice) then
                        choice = tonumber(choice)
                        key_save(db.path.stm32f1.wdg.flags, "__WDG_CLK_DIVIDER__",  divtab[choice][2])
                        key_save(db.path.stm32f1.wdg.flags, "__WDG_RELOAD_VALUE__", divtab[choice][3])
                end

                return choice
        end

        local function configure_behaviour_on_debug()
                local choice = key_read(db.path.stm32f1.wdg.flags, "__WDG_DISABLE_ON_DEBUG__")
                msg(progress().."Do you want to disable WDG on debug?")
                msg("Current selection is: "..filter_yes_no(choice)..".")
                add_item(yes, "Yes")
                add_item(no, "No")
                choice = get_selection()
                if can_be_saved(choice) then
                        key_save(db.path.stm32f1.wdg.flags, "__WDG_DISABLE_ON_DEBUG__", choice)
                end

                return choice
        end

        local function print_summary()
                local dev_lock = key_read(db.path.stm32f1.wdg.flags, "__WDG_DEVICE_LOCK_AT_OPEN__")
                local divider  = key_read(db.path.stm32f1.wdg.flags, "__WDG_CLK_DIVIDER__")
                local reload   = key_read(db.path.stm32f1.wdg.flags, "__WDG_RELOAD_VALUE__")
                local debug    = key_read(db.path.stm32f1.wdg.flags, "__WDG_DISABLE_ON_DEBUG__")
                local lsi      = pll:get_frequencies().LSI

                msg("WDG module configuration summary:")
                msg("Watchdog locked when device open: "..filter_yes_no(dev_lock).."\n"..
                    "Timeout: "..tunit(round(1/(lsi/divider/reload), 3)).."\n"..
                    "Disable on debug: "..filter_yes_no(dev_lock))

                pause()
        end

        if key_read(db.path.project.mk, "ENABLE_WDG") == yes then

                local pages = {configure_dev_lock_at_open, configure_wdg_timeout, configure_behaviour_on_debug, print_summary}
                return show_pages(pages)
        end

        return next
end

--------------------------------------------------------------------------------
-- METHODS
--------------------------------------------------------------------------------
--------------------------------------------------------------------------------
-- @brief Function execute configuration
--------------------------------------------------------------------------------
function wdg:configure()
        title("WDG module configuration for " .. db:get_cpu_name())
        navigation("Home/Modules/WDG")
        calculate_total_steps()

        ::eth_enable::
        if ask_for_enable() == back then
                return back
        end

        if key_read(db.path.project.mk, "ENABLE_WDG") == yes then
                if arch[db:get_arch()].configure() == back then
                        goto eth_enable
                end
        end

        return next
end

-- started without master file
if (master ~= true) then
        show_no_master_info()
end

--------------------------------------------------------------------------------
-- END OF FILE
--------------------------------------------------------------------------------

File: src/system/drivers/wdg/stm32fx/wdg.c
/*==============================================================================
File     wdg.c

Author   Daniel Zorychta

Brief    Watchdog driver.

         Copyright (C) 2017 Daniel Zorychta <daniel.zorychta@gmail.com>

         This program is free software; you can redistribute it and/or modify
         it under the terms of the GNU General Public License as published by
         the  Free Software  Foundation;  either version 2 of the License, or
         any later version.

         This  program  is  distributed  in the hope that  it will be useful,
         but  WITHOUT  ANY  WARRANTY;  without  even  the implied warranty of
         MERCHANTABILITY  or  FITNESS  FOR  A  PARTICULAR  PURPOSE.  See  the
         GNU General Public License for more details.

         You  should  have received a copy  of the GNU General Public License
         along  with  this  program;  if not,  write  to  the  Free  Software
         Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


==============================================================================*/

/*==============================================================================
  Include files
==============================================================================*/
#include "drivers/driver.h"
#include "stm32fx/wdg_cfg.h"
#include "../wdg_ioctl.h"

#if defined(ARCH_stm32f1)
#include "stm32f1/stm32f10x.h"
#define DBGMCU_IWDG_STOP        DBGMCU_CR_DBG_IWDG_STOP
#elif defined(ARCH_stm32f4)
#include "stm32f4/stm32f4xx.h"
#define DBGMCU_IWDG_STOP        DBGMCU_APB1_FZ_DBG_IWDG_STOP
#elif defined(ARCH_stm32f7)
#include "stm32f7/stm32f7xx.h"
#define DBGMCU_IWDG_STOP        DBGMCU_APB1_FZ_DBG_IWDG_STOP
#endif

/*==============================================================================
  Local macros
==============================================================================*/
#define KEY_START               0xCCCC
#define KEY_RELOAD              0xAAAA
#define KEY_UNLOCK              0x5555

/*==============================================================================
  Local object types
==============================================================================*/
typedef struct {
        dev_lock_t file_lock;
} WDG_t;

/*==============================================================================
  Local function prototypes
==============================================================================*/
static inline void configure_wdg(void);
static inline void start_wdg(void);
static inline void reset_wdg(void);

/*==============================================================================
  Local objects
==============================================================================*/
MODULE_NAME(WDG);

/*==============================================================================
  Exported objects
==============================================================================*/

/*==============================================================================
  Function definitions
==============================================================================*/

//==============================================================================
/**
 * @brief Initialize device
 *
 * @param[out]          **device_handle        device allocated memory
 * @param[in ]            major                major device number
 * @param[in ]            minor                minor device number
 * @param[in ]            config               optional module configuration
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_INIT(WDG, void **device_handle, u8_t major, u8_t minor, const void *config)
{
        UNUSED_ARG1(config);

        if (major == 0 && minor == 0) {

                int err = sys_zalloc(sizeof(WDG_t), device_handle);
                if (err == ESUCC) {
                        configure_wdg();
                        start_wdg();
                        reset_wdg();
                }

                return err;

        } else {
                return ENODEV;
        }
}

//==============================================================================
/**
 * @brief Release device
 *
 * @param[in ]          *device_handle          device allocated memory
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_RELEASE(WDG, void *device_handle)
{
        UNUSED_ARG1(device_handle);

        return EPERM;
}

//==============================================================================
/**
 * @brief Open device
 *
 * @param[in ]          *device_handle          device allocated memory
 * @param[in ]           flags                  file operation flags (O_RDONLY, O_WRONLY, O_RDWR)
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_OPEN(WDG, void *device_handle, u32_t flags)
{
        UNUSED_ARG1(flags);

        WDG_t *hdl = device_handle;

        if (_WDG_CFG_OPEN_LOCK) {
                return sys_device_lock(&hdl->file_lock);
        } else {
                return ESUCC;
        }
}

//==============================================================================
/**
 * @brief Close device
 *
 * @param[in ]          *device_handle          device allocated memory
 * @param[in ]           force                  device force close (true)
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_CLOSE(WDG, void *device_handle, bool force)
{
        WDG_t *hdl = device_handle;

        if (_WDG_CFG_OPEN_LOCK) {
                return sys_device_unlock(&hdl->file_lock, force);
        } else {
                return ESUCC;
        }
}

//==============================================================================
/**
 * @brief Write data to device
 *
 * @param[in ]          *device_handle          device allocated memory
 * @param[in ]          *src                    data source
 * @param[in ]           count                  number of bytes to write
 * @param[in ][out]     *fpos                   file position
 * @param[out]          *wrcnt                  number of written bytes
 * @param[in ]           fattr                  file attributes
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_WRITE(WDG,
              void             *device_handle,
              const u8_t       *src,
              size_t            count,
              fpos_t           *fpos,
              size_t           *wrcnt,
              struct vfs_fattr  fattr)
{
        UNUSED_ARG6(device_handle, src, count, fpos, wrcnt, fattr);
        return ENOTSUP;
}

//==============================================================================
/**
 * @brief Read data from device
 *
 * @param[in ]          *device_handle          device allocated memory
 * @param[out]          *dst                    data destination
 * @param[in ]           count                  number of bytes to read
 * @param[in ][out]     *fpos                   file position
 * @param[out]          *rdcnt                  number of read bytes
 * @param[in ]           fattr                  file attributes
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_READ(WDG,
             void            *device_handle,
             u8_t            *dst,
             size_t           count,
             fpos_t          *fpos,
             size_t          *rdcnt,
             struct vfs_fattr fattr)
{
        UNUSED_ARG6(device_handle, dst, count, fpos, rdcnt, fattr);
        return ENOTSUP;
}

//==============================================================================
/**
 * @brief IO control
 *
 * @param[in ]          *device_handle          device allocated memory
 * @param[in ]           request                request
 * @param[in ][out]     *arg                    request's argument
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_IOCTL(WDG, void *device_handle, int request, void *arg)
{
        UNUSED_ARG1(arg);

        WDG_t *hdl = device_handle;

        int err = sys_device_get_access(&hdl->file_lock);
        if (!err) {
                switch (request) {
                case IOCTL_WDG__RESET:
                        reset_wdg();
                        break;

                default:
                        return EBADRQC;
                }
        }

        return err;
}

//==============================================================================
/**
 * @brief Flush device
 *
 * @param[in ]          *device_handle          device allocated memory
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_FLUSH(WDG, void *device_handle)
{
        UNUSED_ARG1(device_handle);

        return ESUCC;
}

//==============================================================================
/**
 * @brief Device information
 *
 * @param[in ]          *device_handle          device allocated memory
 * @param[out]          *device_stat            device status
 *
 * @return One of errno value (errno.h)
 */
//==============================================================================
API_MOD_STAT(WDG, void *device_handle, struct vfs_dev_stat *device_stat)
{
        UNUSED_ARG1(device_handle);

        device_stat->st_major = 0;
        device_stat->st_minor = 0;
        device_stat->st_size  = 0;

        return ESUCC;
}

//==============================================================================
/**
 * @brief Configure Watchdog
 */
//==============================================================================
static inline void configure_wdg(void)
{
        /* default LSI / 4 */
        u8_t divider;
        switch (_WDG_CFG_DIVIDER) {
        case 8  : divider = IWDG_PR_PR_0; break;
        case 16 : divider = IWDG_PR_PR_1; break;
        case 32 : divider = IWDG_PR_PR_1 | IWDG_PR_PR_0; break;
        case 64 : divider = IWDG_PR_PR_2; break;
        case 128: divider = IWDG_PR_PR_2 | IWDG_PR_PR_0; break;
        case 256: divider = IWDG_PR_PR_2 | IWDG_PR_PR_1; break;
        default : divider = 0; break;
        }

        /* reload value */
        u16_t reload = _WDG_CFG_RELOAD & 0xFFF;

        /* enable IWDG */
        IWDG->KR = KEY_UNLOCK;

        while (IWDG->SR & IWDG_SR_PVU);
        IWDG->PR = divider;

        while (IWDG->SR & IWDG_SR_RVU);
        IWDG->RLR = reload;

        if (_WDG_CFG_DISABLE_ON_DEBUG) {
                SET_BIT(DBGMCU->CR, DBGMCU_IWDG_STOP);
        }
}

//==============================================================================
/**
 * @brief Start Watchdog
 */
//==============================================================================
static inline void start_wdg(void)
{
        IWDG->KR = KEY_START;
}

//==============================================================================
/**
 * @brief Reload Watchdog
 */
//==============================================================================
static inline void reset_wdg(void)
{
        IWDG->KR = KEY_RELOAD;
}

/*==============================================================================
  End of file
==============================================================================*/
